lnwallet/chancloser/rbf_coop_transitions.go

```diff
@@ -0,0 +1,1117 @@
+		// last round?
+		if closeTerms.LocalCanPayFees(absoluteFee) {
+			// Each time we go into this negotiation flow, we'll
+			// kick off our local state with a new close attempt.
+			// So we'll emit a internal event to drive forward that
+			// part of the state.
+			localOfferSign := ProtocolEvent(&SendOfferEvent{
+				TargetFeeRate: idealFeeRate,
+			})
+			internalEvents = append(internalEvents, localOfferSign)
+		} else {
+			chancloserLog.Infof("ChannelPoint(%v): unable to pay "+
+				"fees with local balance, skipping "+
+				"closing_complete", env.ChanPoint)
+		}
+
+		if len(internalEvents) > 0 {
+			newEvents = fn.Some(protofsm.EmittedEvent[ProtocolEvent]{
+				InternalEvent: fn.Some(internalEvents),
+			})
+		}
+
+		return &CloseStateTransition{
+			NextState: &ClosingNegotiation{
+				PeerState: DualPeerState{
+					LocalState: &LocalCloseStart{
+						CloseChannelTerms: closeTerms,
+					},
+					RemoteState: &RemoteCloseStart{
+						CloseChannelTerms: closeTerms,
+					},
+				},
+			},
+			NewEvents: newEvents,
+		}, nil
+
+	default:
+		return nil, fmt.Errorf("%w: received %T while in ChannelFlushing",
+			ErrInvalidStateTransition, msg)
+	}
+}
+
+// ProcessEvent drives forward the composite states for the local and remote
+// party in response to new events. From this state, we'll continue to drive
+// forward the local and remote states until we arrive at the StateFin stage,
+// or we loop back up to the ShutdownPending state.
+func (c *ClosingNegotiation) ProcessEvent(event ProtocolEvent, env *Environment,
+) (*CloseStateTransition, error) {
+
+	// There're two classes of events that can break us out of this state:
+	// we receive a confirmation event, or we receive a signal to restart
+	// the co-op close process.
+	switch msg := event.(type) {
+	// If we get a confirmation, then the spend request we issued when we
+	// were leaving the ChannelFlushing state has been confirmed.  We'll
+	// now transition to the StateFin state.
+	case *SpendEvent:
+		return &CloseStateTransition{
+			NextState: &CloseFin{
+				transitionEvent: msg,
+				ConfirmedTx:     msg.Tx,
+			},
+		}, nil
+
+	// Otherwise, if we receive a shutdown, or receive an event to send a
+	// shutdown, then we'll go back up to the ChannelActive state, and have
+	// it handle this event by emitting an internal event.
+	//
+	// TODO(roasbeef): both will have fee rate specified, so ok?
+	case *ShutdownReceived, *SendShutdown:
+		chancloserLog.Infof("ChannelPoint(%v): RBF case triggered, "+
+			"restarting negotiation", env.ChanPoint)
+
+		return &CloseStateTransition{
+			NextState: &ChannelActive{},
+			NewEvents: fn.Some(protofsm.EmittedEvent[ProtocolEvent]{
+				InternalEvent: fn.Some([]ProtocolEvent{event}),
+			}),
+		}, nil
+	}
+
+	// If we get to this point, then we have an event that'll drive forward
+	// the negotiation process.  Based on the event, we'll figure out which
+	// state we'll be modifying.
+	switch {
```

Comment:
So that's an error, we'll fall through and return that we got an invalid event. Or do you think there's a case where something can be meant for both the local and remote states? The local and remote stats accept a disjoint set of events (see the impl of `ShouldRouteTo`) for them both.